internal/executor/shell.go

```diff
@@ -61,15 +60,16 @@ func ShellCommandsAndWait(
 
 	select {
 	case <-ctx.Done():
-		handler([]byte("\nTimed out!"))
```

Comment:
It feels like a proper approach here would be branch on `ctx.Err()`:

```diff
diff --git a/internal/executor/shell.go b/internal/executor/shell.go
index 4545003..9e419cd 100644
--- a/internal/executor/shell.go
+++ b/internal/executor/shell.go
@@ -61,7 +61,15 @@ func ShellCommandsAndWait(
 
        select {
        case <-ctx.Done():
-               handler([]byte("\nTimed out!"))
+               var message string
+
+               if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+                       message = "Timed out!"
+               } else {
+                       message = fmt.Sprintf("Context terminated unexpectedly: %v", ctx.Err())
+               }
+
+               handler([]byte("\n" + message))
 
                processdumper.Dump()
 
@@ -69,7 +77,7 @@ func ShellCommandsAndWait(
                        handler([]byte(fmt.Sprintf("\nFailed to kill a timed out shell session: %s", err)))
                }
 
-               return cmd, TimeOutError
+               return cmd, ctx.Err()
        case <-done:
                var forcePiperClosure bool
 ```

This will keep the old behavior in place and won't munge the error message.

Reply:
IMO the cause approach is better. It's clear what was the cause. `DeadlineExceeded` might come from somehting else in the future, for example.

Reply:
Plus what kind of error we need to return? `ctx.Err()` or `TimeOutError`?